apps/shell/app.js

```diff
@@ -59,37 +56,10 @@ app.use(process.env.PASSENGER_BASE_URI || '/', router);
 const server = new http.createServer(app);
 const wss = new WebSocket.Server({ noServer: true });
 
-let host_allowlist = new Set;
-if (process.env.OOD_SSHHOST_ALLOWLIST){
-  host_allowlist = new Set(process.env.OOD_SSHHOST_ALLOWLIST.split(':'));
-}
-
-let default_sshhost, first_available_host;
-glob.sync(path.join((process.env.OOD_CLUSTERS || '/etc/ood/config/clusters.d'), '*.y*ml'))
-  .map(yml => yaml.safeLoad(fs.readFileSync(yml)))
-  .filter(config => (config.v2 && config.v2.login && config.v2.login.host) && ! (config.v2 && config.v2.metadata && config.v2.metadata.hidden))
-  .forEach((config) => {
-    let host = config.v2.login.host; //Already did checking above
-    let isDefault = config.v2.login.default;
-    host_allowlist.add(host);
-    if (isDefault) default_sshhost = host;
-    if (!first_available_host) first_available_host = host;
-  });
-
-default_sshhost = process.env.OOD_DEFAULT_SSHHOST || process.env.DEFAULT_SSHHOST || default_sshhost || first_available_host;
-if (default_sshhost) host_allowlist.add(default_sshhost);
-
-function host_and_dir_from_url(url){
-  let match = url.match(host_path_rx), 
-  hostname = null, 
-  directory = null;
-
-  if (match) {
-    hostname = match[1] === "default" ? default_sshhost : match[1];
-    directory = match[2] ? decodeURIComponent(match[2]) : null;
-  }
-  return [hostname, directory];
-}
+let host_allowlist = helpers.generateHostAllowlist(process.env.OOD_SSHHOST_ALLOWLIST);
+let cluster_sshhosts = helpers.generateClusterSshhosts(process.env.OOD_CLUSTERS);
+let default_sshhost = process.env.OOD_DEFAULT_SSHHOST || process.env.DEFAULT_SSHHOST || helpers.generateDefaultSshhost(cluster_sshhosts);
+host_allowlist = helpers.addToHostAllowlist(host_allowlist, cluster_sshhosts, default_sshhost);
```

Comment:
There is still too much here. You should have at most 2 functions here. One to get the host_allowlist, and one to get the default_sshhost. An ideal interface would likely require host_allowlist to become an object with its own methods defined - and at that point you could do host_allowlist.default_host() and then only need one. But a simple solution that would take less time now is to combine the two functions:

    const [host_allowlist, default_sshhost] = helpers.hostAllowlistAndDefaultHost(process.env.OOD_SSHHOST_ALLOWLIST, process.env.OOD_CLUSTERS, process.env.OOD_DEFAULT_SSHHOST || process.env.DEFAULT_SSHHOST);

Reply:
Well actually you could just create host_allowlist and then define a default_sshhost function on the host_allowlist object (the function can refer to `this` to call the set methods). Then you end up with something like:

```
let host_allowlist = helpers.generateHostAllowlist(process.env.OOD_SSHHOST_ALLOWLIST, process.env.OOD_CLUSTERS, process.env.OOD_DEFAULT_SSHHOST || process.env.DEFAULT_SSHHOST);

// later on in the code you can call to retrun the default host
host_allowlist.default_sshhost() 
```

Reply:
I will open up a separate PR implementing the class so we can compare the two branches